sympy/matrices/eigen.py

```diff
@@ -19,6 +22,65 @@
 from .utilities import _iszero
 
 
+def _eigenvals_triangular(M, multiple=False):
+    """A fast decision for eigenvalues of an upper or a lower triangular
+    matrix.
+    """
+    diagonal_entries = [M[i, i] for i in range(M.rows)]
+    if multiple:
+        return diagonal_entries
+    return dict(Counter(diagonal_entries))
+
+
+def _eigenvals_eigenvects_mpmath(M):
+    mp = MPContext()
+    prec = max([x._prec for x in M.atoms(Float)])
+
+    mp.prec = prec
+    A = mp.matrix(M.evalf(n=prec_to_dps(prec)))
+    E, ER = mp.eig(A)
+    norm = mp.fsum([mp.re(e)**2 + mp.im(e)**2 for e in E])
+
+    while True:
+        prec_new = round(prec * 1.5)
+
+        if prec_new > DEFAULT_MAXPREC:
+            raise PrecisionExhausted
+
+        mp.prec = prec_new
+        A_new = mp.matrix(M.evalf(n=prec_to_dps(prec)))
+        E_new, ER_new = mp.eig(A_new)
+        norm_new = mp.fsum([mp.re(e)**2 + mp.im(e)**2 for e in E_new])
+
+        if abs(norm - norm_new) < max(norm, norm_new) * 2**-prec:
+            break
+
+        norm = norm_new
```

Comment:
```suggestion
        norm = norm_new
        prec = prec_new
```

Reply:
These changes are needed at least, I think. But see the snippet that I last posted for what I think is a better way to calculate the error. Also, shouldn't we be comparing the difference in the sum of the squares to the square of the error? (Equivalent to comparing the `sqrt` of the sum of the square to the error.)

Reply:
Here the max deviation in the computed values is checked:
```diff
def _eigenvals_eigenvects_mpmath(M):
    from mpmath import workprec
    norm2 = lambda v1, v2: max([max(abs(r1 - r2), abs(i1 - i2))
        for (r1,i1), (r2,i2) in zip(v1, v2)])

    v1 = None
    prec = max([x._prec for x in M.atoms(Float)])
    while prec < DEFAULT_MAXPREC:

        with workprec(prec):
            A = mp.matrix(M.evalf())
            E, ER = mp.eig(A)

        v2 = [(mp.re(e), mp.im(e)) for e in E]
        if v1 is not None:
            if norm2(v1, v2) <  5*2**-(prec//2):
                return E, ER
        v1 = v2

        prec *= 2

    # we get here because the next step would have taken us
    # past MAXPREC or because we never took a step; in case
    # of the latter, we refuse to send back a solution since
    # it would not have been verified; we also resist taking
    # a small step to arrive exactly at MAXPREC since then
    # the norm and new_norm would be artificially close
    raise PrecisionExhausted
```